### DISCUSSION TITLE
One-to-Many Table Relationship

### DISCUSSION THREAD

**Discussion by ghislaineguerin** [mathesar-foundation/mathesar#598] [Maintainer discussions]
The design for issue #243 is ready for review. Please follow the [Review Guidelines](https://wiki.mathesar.org/en/design/process/review-guidelines) and make sure to review before the set deadline expires.

### Document
https://wiki.mathesar.org/en/design/specs/views-one-to-many-relationships

### Reviewers
@kgodey @pavish Your review is required.
Feedback from the community is also welcome. More context about Mathesar's design strategy can be found [here](https://wiki.mathesar.org/design).

### Deadline
The deadline for submitting feedback is set 2 days from now, on Friday 26th.

**Comment by kgodey** [comment]
@ghislaineguerin This looks good. Some feedback:

(1) I think it would be good to display the first (non-primary key) column of the referenced table in the same column, similar to the dropdown for selecting a new record.
<img width="278" alt="Screen Shot 2021-08-25 at 2 18 16 PM" src="https://user-images.githubusercontent.com/287034/130844357-247ba458-fbe4-4794-8626-9d1e0049a4cb.png">

(2) There may be thousands or millions of records in the table that's being referenced. We need a way to link the correct record. We probably need some sort of search and autocomplete while linking records. Search should probably account for other fields in the table, not just the PK (e.g. if I want to link `Aqua` to `Aquarium`, I probably want to start typing `Aq`, not the ID of `Aqua`)

(3) I don't think we need the "From Referenced Tables" subheader in the table constraints section, all constraints are the same from the DB point of view. Can we find a different way to convey what you're trying to convey here? Maybe a tooltip?

(4) I think we might want different icons for primary keys and foreign keys to make the distinction clearer. Maybe some sort of "link" icon for FKs?

**Reply by ghislaineguerin** [reply]
Thanks @kgodey 
1. Agree, I've added to the prototype
2. Correct, I will describe this in more detaili the specs doc
3. I've added a tooltip instead
4. I had created an icon already for the next issue, it references the column and a key

**Reply by kgodey** [reply]
(1) and (3) look good.

I assume (2) is not done yet.

Re: (4), I think having a single separate icon will be cleaner than having two icons combined.

**Reply by ghislaineguerin** [reply]
@kgodey agreed, I wanted to keep the key concept present in the icon, but maybe we can do that without it, I'm concerned that the link icon will be too similar to URL data type. I will think about options.

**Reply by ghislaineguerin** [reply]
@kgodey 
- Autocomplete has been added to the dropdown for adding records
- Updated icons have been added

**Reply by kgodey** [reply]
@ghislaineguerin The autocomplete looks good. The icon is hard to make out for me, I'm not sure what it represents.

**Reply by ghislaineguerin** [reply]
@kgodey I'm adding a note to review icons globally once the milestone design issues are ready.

**Reply by kgodey** [reply]
Sounds good, thanks.

**Comment by mathemancer** [comment]
A couple of things I noticed:
- this is actually describing a one-to-many relationship in general (one-to-one is a special case of that).  one-to-many is the default foreign key relationship, and the most useful.
- The "orphan records" concept breaks the data consistency gained through use of foreign keys (in fact, this can't be easily implemented using foreign keys at the DB level; we'd have to get pretty hacky).  What are we hoping to gain from this?  Is it really necessary?  I'd want to understand the use case more specifically if we are really sure that such a feature is needed:
  - In case of some "orphan records", am I correct that no _new_ records would be added using that "orphan" value?
  - Would it be possible to update a record to use an "orphan" value?
  - When updating a record already containing an "orphan" value, should the new value for that cell be restricted by the foreign key linkage?  What if that cell isn't modified in the update statement?  Should the orphan value remain?

The usual options for what to do when a referenced value is deleted are:
- throw an error (warning) and refuse; this is the default for PostgreSQL.
- delete all rows that reference that value in the other table
- Set `NULL` wherever the deleted value was referenced.
- Set some default value wherever the deleted value was referenced (the default must be a valid id in the referenced table).

Relevant docs:  https://www.postgresql.org/docs/13/ddl-constraints.html#DDL-CONSTRAINTS-FK

**Reply by kgodey** [reply]
For some reason, I was thinking of "orphan records" as "if the foreign key constraint is deleted from this column/if the referenced column is deleted, then keep the column on this table and its values intact". I'm not sure if that was the intent behind the design, though.

**Reply by mathemancer** [reply]
Ah, I was interpreting it as "allow us to keep the value in the referencing table if the value is removed from the referenced table"

More clarification would be great there.

**Reply by ghislaineguerin** [reply]
@mathemancer Yes, it's a one-to-many as I didn't have any good examples to demonstrate a 1 to 1. I agree that it's most useful. 
As for orphan records, I did some research on data integrity and how it might be enforced, my concern was that since we don't have integrated views with the related tables, the user might forget or not know that a table record is referenced elsewhere and might accidentally create 'orphaned table records'.

This is an excerpt explaining how this feature is implemented in Microsoft Access:
> The rules of integrity, by default, are designed to prevent the incidence of orphaned table records on the “many” side of relationships. As such, Microsoft Access prohibits the deletion of parent records or the modification of primary key values when associated child records exist. If referential integrity is not established, an indeterminate relationship exists between tables.

I'm open to ideas on how we can make that obvious to the users, as we won't have lookup fields in tables, only for views.

**Reply by mathemancer** [reply]
I think it would be much better to make the constraints involved in foreign key references obvious to the user, rather than allowing orphan records.  Maybe a helper to make sure they chose an appropriate value?  (drop-down combined with typing the first few characters?)

**Reply by kgodey** [reply]
Since the orphan records concept has been causing confusion to everyone on this thread and will also complicate implementation, I think we should remove the option and just use default foreign key behavior.

**Reply by kgodey** [reply]
Also, perhaps we should rename the spec "one-to-many" to avoid future confusion. I don't think we need to do anything specific for one-to-one at this point, it's just a case of one-to-many.

**Comment by pavish** [comment]
@ghislaineguerin @kgodey  I have a few questions and thoughts:

**In the menu**
* We are linking tables `release` and `artist` through existing an existing column `artist_id`'. What happens when the values in the rows are not present on the referenced table `artist`?
  - This will throw an error from the db, how do we represent it on the frontend?
* I do not understand what 'Orphan records' mean. I assume it means we allow records to exist in `release` table when there is no linkable records on the referenced table `artist`, but the DB will not allow that.
* When clicking the option in 'Select table', it opens a new modal. It should be the same model. I assume you've meant it to be the same modal, but the prototype shows it as such. It is better to update the UX, since we have other contributors who may work on this and get confused.

**The table**
* I think it might be better to ask the user to define lookup fields, like airtable does, for searching records, instead of the first non-primary column. The first non-primary column may be a non-unique column, which would not be useful.
* We should not allow the user to type in values directly to a linked cell. Instead, it should act like a search box. The cell should only be updated when a valid value is clicked from the dropdown, and should clear if clicked elsewhere.
* While deleting a record from a referenced table (`artist`), we can show the tables that are linked, in the alert modal.

We also need to factor in how the UI will look when there is a one-many or many-one relationship involved. Since the menu options and modal is going to be similar for all three, it is better to start development work after we define the UX for all three. Or else, each design iteration will result in a complete rework on the frontend.

**Reply by ghislaineguerin** [reply]
Thanks @pavish 

> We are linking tables release and artist through existing an existing column artist_id'. What happens when the values in the rows are not present on the referenced table artist?
> This will throw an error from the DB, how do we represent it on the frontend?

I think we need to define this for other scenarios as well, for example, if we have dates as strings and we transform, what do we do with the non parseable values. I think we should either clear those fields or prevent the change.
@kgodey is this something we need to include in global data types as well?

> I do not understand what 'Orphan records' mean. I assume it means we allow records to exist in the release table when there are no linkable records on the referenced table artist, but the DB will not allow that.

It's a record containing a foreign key that is linking to a deleted parent table record. So in this case, a record in releases was pointing to artist with id '34' and now we delete that artist from the 'artists' table

> When clicking the option in 'Select table,' it opens a new modal. It should be the same model. I assume you've meant it to be the same modal, but the prototype shows it as such. It is better to update the UX since we have other contributors who may work on this and get confused.

Yes, I will fix this. It was a limitation with Figma.

> I think it might be better to ask the user to define lookup fields, like airtable does, for searching records, instead of the first non-primary column. The first non-primary column may be a non-unique column, which would not be useful.

Airtable does that from a view, not the actual tables as they are "hidden". The idea we've discussed w @kgodey is that columns from foreign key tables can only be added from views. 

> We should not allow the user to type in values directly to a linked cell. Instead, it should act as a search box. The cell should only be updated when a valid value is clicked from the dropdown and should clear if clicked elsewhere.

We'd have to define how they paste values into these cells. But I think you've got the right idea here, although it would be nice that they can just start typing and trigger the autocomplete.

> While deleting a record from a referenced table (artist), we can show the tables that are linked in the alert modal.

Makes sense. I will update.

> We also need to factor in how the UI will look when there is a one-many or many-one relationship involved. Since the menu options and modal is going to be similar for all three, it is better to start development work after we define the UX for all three. Or else, each design iteration will result in a complete rework on the front end.

Agreed.

**Reply by kgodey** [reply]
> I think we need to define this for other scenarios as well, for example, if we have dates as strings and we transform, what do we do with the non parseable values. I think we should either clear those fields or prevent the change.
@kgodey is this something we need to include in global data types as well?

Currently we prevent the change (to avoid data loss) so we should show an error message. We should include this in global data types too. We can reevaluate whether to clear fields when we have undo/redo.

> It's a record containing a foreign key that is linking to a deleted parent table record. So in this case, a record in releases was pointing to artist with id '34' and now we delete that artist from the 'artists' table

See my comment above regarding orphan records: https://github.com/centerofci/mathesar/discussions/598#discussioncomment-1264315

> I think it might be better to ask the user to define lookup fields, like airtable does, for searching records, instead of the first non-primary column. The first non-primary column may be a non-unique column, which would not be useful.

I think @pavish is suggesting that the user can designate a field in the table as a "lookup field", and we would use that field instead of the first field to identify the record. For example, imagine this table:

| id | date of birth | astrological sign | artist name |
|-|-|-|-|
| 1 | 1950-01-01 | Capricorn | Cool Artist |
| 2 | 1979-01-12 | Capricorn | Amazing Artist |

In this case, in our current design, the user would see the date of birth while trying to link an artist to a release, but that would not be useful. Astrological sign would be even less useful (if it was the first column). The user could have an option to designate artist name as a "lookup field"  so that that would be the field shown while trying to find records or in foreign keys. I think this is a good idea.

**Reply by pavish** [reply]
@kgodey has covered most of the points I wanted to reply.

> We'd have to define how they paste values into these cells. But I think you've got the right idea here, although it would be nice that they can just start typing and trigger the autocomplete.

I agree. We need to define the UX for these scenarios in more detail, especially for all invalid states.

**Reply by ghislaineguerin** [reply]
@kgodey @pavish I have updated the prototype to address the following:
- Columns now show a 'lookup field' icon at the column header level. This will be set by default for the first column after ID
- Table actions menu contains an option to set the lookup field if the user wants to change the default
- Adding the column name to the lookup field results is no longer needed, only the values of the lookup field are displayed
- Show affected tables when deleting a linked record

Question:
- Being a foreign key field, I assume the user can always enter the exact ID number and also link the record (no need to know the lookup field values)?

**Reply by kgodey** [reply]
> Columns now show a 'lookup field' icon at the column header level. This will be set by default for the first column after ID
> Table actions menu contains an option to set the lookup field if the user wants to change the default

Looks good! I think we need some place to explain what a "lookup field" is in the modal where we set it.

> Adding the column name to the lookup field results is no longer needed, only the values of the lookup field are displayed

I'm not sure what this means. Do you mean you don't show "artist_name" before the value?

I think we need some sort of helper when linking a table that says "We're using the "X" column to look up values. If these values not helpful? Please set a different lookup field on the table here: <button>" since the user may not have set a lookup field before deciding to link the table.

> Show affected tables when deleting a linked record

This looks good. I think it would be nice to provide a button to go to each table that is referenced and to show the column in the table that is referencing this record.

Also, this is a separate thought, but I think for columns with foreign key constraints, it would be good to have an option to "go to the referenced table" in the column (similar to views).

> Being a foreign key field, I assume the user can always enter the exact ID number and also link the record (no need to know the lookup field values)?

Yes, that's correct.

**Reply by ghislaineguerin** [reply]
> Looks good! I think we need some place to explain what a "lookup field" is in the modal where we set it.

Makes sense. I've added one.

> I think we need some sort of helper when linking a table that says "We're using the "X" column to look up values. If these values not helpful? Please set a different lookup field on the table here: " since the user may not have set a lookup field before deciding to link the table

Agreed. When you say 'here,' do you mean we can link to the actual table? If yes, would that open the referenced table in a new tab?

> This looks good. I think it would be nice to provide a button to go to each table that is referenced and to show the column in the table that is referencing this record.

I'm adding this already for 'Creating, Editing, deleting views'

**Reply by kgodey** [reply]
> Makes sense. I've added one.

I don't see any explanation in the modal, here's what I see:
<img width="621" alt="Screen Shot 2021-09-06 at 8 44 39 AM" src="https://user-images.githubusercontent.com/287034/132219518-962c4ae6-639f-48d4-bde9-6f55e835a287.png">
 
> Agreed. When you say 'here,' do you mean we can link to the actual table? If yes, would that open the referenced table in a new tab?

Yes to both questions. I assume we'd also switch to the new tab.

**Reply by ghislaineguerin** [reply]
@kgodey I've added the explanation to the modal where the table is selected, the modal you are showing here is where the lookup column is set on the actual table. Let me know if this is what you meant.

![image](https://user-images.githubusercontent.com/845767/132407353-53e990d0-8e36-442c-826d-7c45f36e2773.png)

**Reply by kgodey** [reply]
@ghislaineguerin I think we need an explanation in the modal that I screenshotted as well, people aren't going to know what a lookup field is.

Also, the explanation that you set up doesn't actually explain what data lookup means.

**Reply by ghislaineguerin** [reply]
@kgodey got it, I will work on the message for both modals.

**Reply by ghislaineguerin** [reply]
@kgodey a new message describing lookups can be found in the most recent versions of the prototype. Is it ok to address this then?

**Reply by kgodey** [reply]
@ghislaineguerin Sure, please add a note to the spec where it should be reviewed so that I don't forget to review it.

**Comment by kgodey** [comment]
@mathemancer @pavish @ghislaineguerin FYI I renamed the spec in the wiki to be "one-to-many" and updated the link in the description.

**Comment by kgodey** [comment]
Based on the new action plan from the [2021-10-15 meeting](https://wiki.mathesar.org/meeting-notes/october-2021), we're going to update the design requirements for issues in this milestone, including this one.
